File: Java/interview_31.java
class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        Stack<Integer> stack = new Stack<>();
        int popIdx = 0;
        for (int p : pushed) {
            if (p == popped[popIdx]) {
                ++popIdx;
                while (stack.size() > 0 && stack.peek() == popped[popIdx]) {
                    stack.pop();
                    ++popIdx;
                }
            }
            else {
                stack.push(p);
            }
        }
        return stack.size() == 0;
    }
}

File: python/interview_31.py
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        popIdx = 0
        stack = []
        for p in pushed:
            if p == popped[popIdx]:
                popIdx += 1
                while len(stack) > 0 and stack[-1] == popped[popIdx]:
                    stack.pop()
                    popIdx += 1
            else:
                stack.append(p)
        return len(stack) == 0
